Reimplement toplevel typechecking with trampoline macros

takikawa opened the pull request:
This pull request reimplements top-level typechecking. This new approach fixes the PR13747 regression due to the new expander and lets us eliminate hacks used to support the top-level. It is an alternative to PR #172.

One slight backwards incompatibility is that I changed `:print-type` and related functions to only work on expressions and not definitions or other top-level only constructs such as `require/typed`. I doubt people ever use `:print-type` on anything but expressions, so this seems ok.

I'll squash most of these commits before pushing.

samth commented on typed-racket-lib/typed-racket/base-env/top-interaction.rkt:
```diff
@@ -77,16 +77,16 @@
+       (define type
+         (tc-expr (local-expand #'e 'expression (list #'module*))))
+       #`(displayln
+          #,(if (eq? type 'no-type)
```
Does this happen anymore given your changes?

takikawa replied:
Oh good catch, no I don't think it does.

stamourv commented:
I'd consider `:print-type` on non-expressions to be a bug, so that's fine by me.

samth commented on typed-racket-lib/typed-racket/base-env/top-interaction.rkt:
```diff
@@ -114,17 +114,17 @@
     (syntax-parse stx
```
I feel like these three forms share a lot of code ...

takikawa commented:
Ok, I've addressed the code feedback so far.

I did abstract the `#%top-interaction` forms, but I'm not sure it actually makes the code easier to read. I pushed the commit anyway, but what do you think?
